### DISCUSSION TITLE
[Person] Different ages for a specific search

### DISCUSSION THREAD

**Discussion by AllieLerbs** [SORMAS-Foundation/SORMAS-Project#11828] [Issue Discussions]
### Bug Description

If you set date filters for person in person directory the age of persons get displayed differently.

### Reproduction Steps

1. Login into SORMAS
2. Go to menu "Persons"
3. Enter "1989" in year of birth search field
4. Enter "May" in month of birth search field
5. Enter "6" in day search field
6. Click on button "Apply filters"
7. Check column "Age and birth of date"
8. If you click on person ID the person opens and the correct age is visible

### Expected Behavior

If you search for a specific date the age of the person the age have to be displayed correctly.

### Affected Area(s)

Web app

### SORMAS Version

SORMAS 1.83.0-Snapshot (70a685d)

### Android version/Browser

Edge

### User Role(s)

Admin

### Server URL

https://test-auto.sormas.netzlink.com/

### Additional Information

![image](https://user-images.githubusercontent.com/91321821/229772698-1c735f4f-f9a7-4d41-b663-0dd044e425d4.png)

![image](https://user-images.githubusercontent.com/91321821/229772782-25e27af4-9f1d-493c-88ab-446da217663c.png)

**Comment by AllieLerbs** [comment]
Possible issue can result in the fact that the field is stated with "approximateAge" and there is no correct counting of the correct age.

**Comment by sergiupacurariu** [comment]
@AliciaRaber @bernardsilenou @MateStrysewske 
This difference comes from the flexibility with which the age and birthday fields are implemented.
First of all, let’s walk briefly through the process of adding the birthdate. This can be done in two ways:
-	By adding the birth date of the person. Once the birth year is added the age field is no longer available. It will be calculated.
-	By adding the age of the person. It is possible to add only the age without adding anything to the birth date fields.
In order to accommodate this approach both birth date and age are stored separately in the database. The age calculation logic occurs on person details loading or if the birth year is updated. When the person's age is displayed in the person grid the data is picked from the database without any additional calculation or check.
The following situations arrive from this approach:
-	If the user fills only the age field (no birth year) and it passes more than a year since the age was added the age remains the same. At this point, there is no automatic process to do this 
-	If the user fills the birth year field then the person's age is calculated automatically, filled in the age field and the field is made read-only. Birth month and day do not count if there is no birth year. On person save the birth year and the person's age are saved in separate fields. The age field is not updated automatically so if more than a year passes the age will become obsolete.

Below you have my thoughts as a starting point:

1. Create a CronJob which updates the person's age periodically using "Approximate age reference date"
2.    - Remove the age field from the database and convert the age inserted by the user to an actual year
       - Calculate the age during person grid loading, change sorting from "Approximate age" to Birth date

**Comment by MateStrysewske** [comment]
@sergiupacurariu Is this really relevant here? Because all of these persons have a concrete birthdate, including year, month, and day, so the automatic age calculation shouldn't really have an effect here, should it? As @AliciaRaber mentioned, in the person pages themselves the age is always calculated correctly, so it's supposed to be a display issue in the person directory, right?

**Comment by sergiupacurariu** [comment]
@MateStrysewske Indeed initially I thought the same a display issue, but after some investigation, it turned to be something else. In the person detail page the age is shown correctly because this calculation is done in toPersonDto when the person page is loaded no matter what is stored in the database. If the user would just saves the person without doing anything else then the newly calculated age will be stored in the database. If the user does this then the grid will show the correct age.

**Comment by bernardsilenou** [comment]
@MateStrysewske @sergiupacurariu 
- The computation of age can depend on the entity referenced. The age (through the case view) is generally defined as the age at the time (report date) that the person was a case. But when looking at the general person entity, then we show the age of the personal based on the current calendar date (current implementation https://github.com/hzi-braunschweig/SORMAS-Project/issues/11788#issuecomment-1503207768). 
- 
- What about have two types of methods to compute age?
- It's generally not a big issue if the age is not updated. The important data is the birth date.
- Will adding a cronjob not lead to performance issues? There are instances with over 100 million persons.

**Comment by MateStrysewske** [comment]
@bernardsilenou @AliciaRaber @sergiupacurariu I've converted the issue into a discussion. Since it originates from the specifications concerning how ages are calculated, I've also turned it into a change request and moved it out of the current iteration because we need to find a proper solution here, which is not trivial.

I'll have another look at it once I find the time; I agree however that a cronjob might be problematic in terms of performance.
